### PR TITLE
[ENH] Fcn pytorch migration

### PR DESCRIPTION
**Pull request by ali-john**
#### Reference Issues/PRs
Adds new FCN classifier in PyTorch as part of https://github.com/sktime/sktime/issues/8699

#### What does this implement/fix? Explain your changes.

- Implemented network structure for FCN in PyTorch

- Implemented FCN classifier in PyTorch
- For files which have both a TensorFlow & a PyTorch version, I have followed the directory/naming structure suggested by maintainers [here](https://github.com/sktime/sktime/pull/8712#pullrequestreview-3132971339.)

#### Does your contribution introduce a new dependency? If yes, which one?
No.

#### What should a reviewer concentrate their feedback on?
- Use of the PyTorch library in the methods that build and train the model.
- parameters that are passed into functions, the new library uses to construct this model.
- to check similar behaviour as used, as in TensorFlow counterpart.

#### Did you add any tests for the change?
- No. The new classifier should be automatically tested by TestAllClassifiers.

#### Any other comments?

No.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://github.com/all-contributors/all-contributors/blob/master/docs/emoji-key.md)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [x] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

File: sktime/classification/deep_learning/fcn/_fcn_torch.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    """Fully Convolutional Network (FCN) for time series classification in PyTorch.
+
+    Adapted from the implementation of Fawaz et. al
+    https://github.com/hfawaz/dl-4-tsc/blob/master/classifiers/fcn.py
+
+    For a drop-in replacement based on TensorFlow, see `FCNClassifier`.
+
+    Parameters
+    ----------
+    activation : str or None or an instance of activation functions defined in
```
Comment (RecreationalMath) [inline sktime/classification/deep_learning/fcn/_fcn_torch.py:31]: In sktime, we have distinction between activations to be used within the hidden layers and the activation (if any) to be used in the output layer via two parameters `activation` and `activation_hidden`, but this implementation seems to implement only a single parameter `activation` and uses it at both places. 

Check here: https://github.com/sktime/sktime/blob/4aab650ca9ab09d76b2b56e0c26b6a26d8a0c640/sktime/classification/deep_learning/rnn/_rnn_torch.py#L26-L32

File: sktime/networks/fcn/_fcn_torch.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+            )
+        
+        nnConv1d = _safe_import("torch.nn.Conv1d")
+        nnBatchNorm1d = _safe_import("torch.nn.BatchNorm1d")
+        nnReLU = _safe_import("torch.nn.ReLU")
+        nnAdaptiveAvgPool1d = _safe_import("torch.nn.AdaptiveAvgPool1d")
+        nnLinear = _safe_import("torch.nn.Linear")
+
+        # activation
+        self.relu = nnReLU()
```
Comment (RecreationalMath) [inline sktime/networks/fcn/_fcn_torch.py]: Activations should be used via the defined parameter `activation` which is in turn stored in `self.activation`, `self.relu` is non-conforming to the overall design.

Comment (RecreationalMath) [reply sktime/networks/fcn/_fcn_torch.py]: Also, I don't see`self.relu` getting used anywhere in the class.

File: sktime/networks/fcn/_fcn_torch.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        self.conv1 = nnConv1d(in_channels=input_features, out_channels=128, kernel_size=8, padding='same')
+        self.bn1 = nnBatchNorm1d(num_features=128)
+
+        # convolutional block 2: 256 filters, kernel size 5
+        self.conv2 = nnConv1d(in_channels=128, out_channels=256, kernel_size=5, padding='same')
+        self.bn2 = nnBatchNorm1d(num_features=256)
+
+        # convolutional block 3: 128 filters, kernel size 3
+        self.conv3 = nnConv1d(in_channels=256, out_channels=128, kernel_size=3, padding='same')
+        self.bn3 = nnBatchNorm1d(num_features=128)
```
Comment (RecreationalMath) [inline sktime/networks/fcn/_fcn_torch.py]: I see that you have hard-coded the number of layers to be 3. Ideally it should be exposed to the end-user via a parameter and a meaningful default or a default borrowed from paper in the references should be used.

File: sktime/networks/fcn/_fcn_torch.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        nnAdaptiveAvgPool1d = _safe_import("torch.nn.AdaptiveAvgPool1d")
+        nnLinear = _safe_import("torch.nn.Linear")
+
+        # activation
+        self.relu = nnReLU()
+        # global average pooling layer
+        self.gap = nnAdaptiveAvgPool1d(1)
+
+        # convolutional block 1: 128 filters, kernel size 8
+        self.conv1 = nnConv1d(in_channels=input_features, out_channels=128, kernel_size=8, padding='same')
```
Comment (RecreationalMath) [inline sktime/networks/fcn/_fcn_torch.py]: All the arguments being passed to convolutional layer are important hyper-parameters of the model and should be exposed to the end-user or infered from the input being passed, instead of hard-coded. 
This is important because we want the model to be easily modifiable as per the end-user's task and specification. In principle, user should always be able to change the model by changing the hyper-parameters.

File: sktime/networks/fcn/_fcn_torch.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        self.bn1 = nnBatchNorm1d(num_features=128)
```
Comment (RecreationalMath) [inline sktime/networks/fcn/_fcn_torch.py]: Same as hyper-parameters for convolutional layers. These should be either exposed to the end-user or infered from the input being passed, not hard-coded.

File: sktime/networks/fcn/_fcn_torch.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        # convolutional block 2: 256 filters, kernel size 5
+        self.conv2 = nnConv1d(in_channels=128, out_channels=256, kernel_size=5, padding='same')
+        self.bn2 = nnBatchNorm1d(num_features=256)
+
+        # convolutional block 3: 128 filters, kernel size 3
+        self.conv3 = nnConv1d(in_channels=256, out_channels=128, kernel_size=3, padding='same')
+        self.bn3 = nnBatchNorm1d(num_features=128)
+
+        # fully connected layer
+        self.fc = nnLinear(in_features=128, out_features=self.num_classes)
```
Comment (RecreationalMath) [inline sktime/networks/fcn/_fcn_torch.py]: The current design for deep-learning classifiers is to divide the overall model into two components, all the layers that come before the fully-connected output layer and fully-connected output layer. 
May I suggest using a oontainer to wrap around all the layers that come before the fully-connected output layer.

File: sktime/networks/fcn/_fcn_torch.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        out = self.bn1(out)
+        out = self.relu(out)
+
+        out = self.conv2(out)
+        out = self.bn2(out)
+        out = self.relu(out)
+
+        out = self.conv3(out)
+        out = self.bn3(out)
+        out = self.relu(out)
```
Comment (RecreationalMath) [inline sktime/networks/fcn/_fcn_torch.py]: Borrowing from the above comment regarding using a container to store the layers - once that is implemented, forward method won't need to specify each layers' operation individually.

**Review by RecreationalMath — COMMENTED** [review]
I have left some structural comments, in order for this estimator to adhere to the design of torch based deep learning estimators in sktime. Will leave a separate review on the FCN aspects of the estimator once these are implemented.

Nice job overall! 👍

@fkiraly Could you please initiate the test run on this PR?

File: sktime/classification/deep_learning/fcn/_fcn_tf.py [hunk 8]
```diff
@@ -17,6 +17,7 @@ class FCNClassifier(BaseDeepClassifier):
     Adapted from the implementation from Fawaz et. al
     https://github.com/hfawaz/dl-4-tsc/blob/master/classifiers/fcn.py
 
+
```
Comment (fkiraly) [inline sktime/classification/deep_learning/fcn/_fcn_tf.py]: no need to add this newline

**Comment by ali-john** [comment]
@RecreationalMath Thanks for this feedback. I will incorporate this. 
I will also look into why the tests are failing.

**Comment by RecreationalMath** [comment]
> @RecreationalMath Thanks for this feedback. I will incorporate this. I will also look into why the tests are failing.

[This guide](https://www.sktime.net/en/stable/developer_guide/coding_standards.html#setting-up-local-code-quality-checks) will help with fixing the code-quality test failures.

**Comment by ali-john** [comment]
@fkiraly , @RecreationalMath  Can you please review it now?

File: sktime/classification/deep_learning/fcn/_fcn_torch.py [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+    n_conv_layers : int, default = 3
+        Number of convolutional blocks in the network
+    filter_sizes : list of int or None, default = None
+        Number of filters for each convolutional layer. If None, defaults to
+        [128, 256, 128] for 3 layers as specified in the original paper
+    kernel_sizes : list of int or None, default = None
+        Kernel sizes for each convolutional layer. If None, defaults to [8, 5, 3]
+        for 3 layers as specified in the original paper
+    activation : str or None or an instance of activation functions defined in
+        torch.nn, default = None
```
Comment (RecreationalMath) [inline sktime/classification/deep_learning/fcn/_fcn_torch.py]: While setting default values, a couple things need to be checked:
1. If there is a reference research paper present in the docstring then defaults should be set according to that paper, so that by default the model behaves as per the settings defined in that research paper. For `FCN` there are two references [Wang et al 2017](https://arxiv.org/pdf/1611.06455) and [Fawaz et al 2019](https://arxiv.org/pdf/1809.04356) (check section 3.2.2 for `FCN`) & the corresponding [Github repo](https://github.com/hfawaz/dl-4-tsc/blob/master/classifiers/fcn.py) for Fawaz et al. As per these references, final layer activation seems to be a `softmax`, same should be used as a default here.
2. Since this is a migration of the model from `TensorFlow` backend to `PyTorch` backend, ideally speaking the same defaults should be used here as are used in the `TensorFlow` implementation. Which turns out to be a `sigmoid` in this case. https://github.com/sktime/sktime/blob/0dbd728952390c63ad979e50fb356412ead3628e/sktime/classification/deep_learning/fcn.py#L35-L35

Since the above two points, mention two different defaults, in my opinion we should stick with the default mentioned in the referenced research papers, i.e. `softmax`.

@fkiraly Do you remember if there was a specific reason for setting the default of final layer `activation` in `TensorFlow` implementation of `FCN` to be `sigmoid` whereas the referenced research papers set it as `softmax`?

File: sktime/classification/deep_learning/fcn/_fcn_torch.py [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+    optimizer_kwargs : dict or None, default = None
+        Additional keyword arguments to pass to the optimizer.
+    criterion : case insensitive str or None or an instance of a loss function
+        defined in PyTorch, default = "CrossEntropyLoss"
+        The loss function to be used in training the neural network.
+        List of available loss functions:
+        https://pytorch.org/docs/stable/nn.html#loss-functions
+    criterion_kwargs : dict or None, default = None
+        Additional keyword arguments to pass to the loss function.
+    callbacks : None or str or a tuple of str, default = None
```
Comment (RecreationalMath) [inline sktime/classification/deep_learning/fcn/_fcn_torch.py:56]: Same as above comment regarding setting defaults.
I understand the tensorflow implementation keeps it as `None` but [Fawaz et al 2019](https://arxiv.org/pdf/1809.04356) (check section 3.2.2 for `FCN`) seems to use `ReduceLROnPlataeu` in their corresponding [Github repo](https://github.com/hfawaz/dl-4-tsc/blob/master/classifiers/fcn.py). Could you please check/confirm this and mention your findings as a comment? Use the priority order as mentioned in the above comment for the final implementation in this PR.

File: sktime/networks/fcn/_fcn_torch.py [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+            )
+            # Add batch norm
+            layers.append(nnBatchNorm1d(num_features=out_channels))
+            # Add activation
+            layers.append(self._instantiate_hidden_activation())
+
+            in_channels = out_channels
+
+        # Add global average pooling
+        layers.append(nnAdaptiveAvgPool1d(1))
```
Comment (RecreationalMath) [inline sktime/networks/fcn/_fcn_torch.py:155]: The above comment mentions Global Average Pooling but the subsequent code implements Adaptive Average Pooling.

File: sktime/networks/fcn/_fcn_torch.py [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+        super().__init__()
+
+        if isinstance(input_size, int):
+            input_features = input_size
+        elif isinstance(input_size, tuple):
+            if len(input_size) == 3:
+                input_features = input_size[1]
+            else:
+                raise ValueError(
+                    "If `input_size` is a tuple, it must either be of length 3 and in "
```
Comment (RecreationalMath) [inline sktime/networks/fcn/_fcn_torch.py]: You can remove the word "either" from the warning message.

File: sktime/networks/fcn/_fcn_torch.py [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+                    "format (n_instances, n_dims, series_length). "
+                    f"Found length of {len(input_size)}"
```
Comment (RecreationalMath) [inline sktime/networks/fcn/_fcn_torch.py]: "Found length of `input_size` to be {len(input_size)}"

File: sktime/classification/deep_learning/fcn/_fcn_torch.py [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+    >>> clf.fit(X_train, y_train) # doctest: +SKIP
+    FCNClassifierTorch(...)
+    """
+
+    _tags = {
+        # packaging info
+        # --------------
+        "authors": ["Ali-John"],
+        "maintainers": ["Ali-John"],
+        "python_version": ">=3.9",
```
Comment (RecreationalMath) [inline sktime/classification/deep_learning/fcn/_fcn_torch.py]: I think sktime now supports only python>=3.10.

https://github.com/sktime/sktime/blob/0dbd728952390c63ad979e50fb356412ead3628e/pyproject.toml#L32

File: sktime/networks/fcn/_fcn_torch.py (hunk 11 above)
Comment (RecreationalMath) [reply sktime/networks/fcn/_fcn_torch.py:155]: I think you mis-understood the fix. 

The comment (Global Average Pooling) was correct and the subsequent code (Adaptive Average Pooling) was incorrect.
The references (Wang et al and Fawaz et al) & the TensorFlow implementation mention a Global Pooling layer not an Adaptive Average Pooling layer. 
https://github.com/sktime/sktime/blob/0dbd728952390c63ad979e50fb356412ead3628e/sktime/networks/fcn.py#L85

Comment (ali-john) [reply sktime/networks/fcn/_fcn_torch.py:155]: I think `nnAdaptiveAvgPool1d(1)` is equivalent to `keras.layers.GlobalAveragePooling1D()`.  What do you suggest?

Comment (RecreationalMath) [reply sktime/networks/fcn/_fcn_torch.py:155]: Ahh, yes, you're right. I confused the Keras and Torch APIs.

In that case do check the dimensionality of the output of the pooling layer. It might need a squeeze operation on the last dimension.
